### ISSUE TITLE
Add self-signed cert support using mkcert backend and GUI option "Create self-signed cert"

### ISSUE THREAD

**Issue by typoworx-de** [jc21/nginx-proxy-manager#593]
**Is your feature request related to a problem? Please describe.**
I'm running docker instance in intranet/private-network and require https for some docker-instances (like docker-registry). The private-lan runs on TLD ".lan" so it's not possible to use letsencrypt by routing domain-name over router as letsencrypt only supports domains/tld's reachable from intranet.

**Describe the solution you'd like**
I noticed nginx-proxy-manager already supports custom-certs which is awesome! I would love to have an additional option in that dropdown in section "SSL Certificates" that could be named "Create self-signed certificate" and then routes this request to f.e. the linux-tool mkcert.

https://blog.filippo.io/mkcert-valid-https-certificates-for-localhost/

```
mkcert my-private-domain.lan
```

**Describe alternatives you've considered**
I could run mkcert on my local machine and manually upload the cert-files into nginx-proxy-manager.

**Additional context**
I think I'm not the only user who runs a docker instance in private/lan and think this feature would support/help other users as well.

**Comment by typoworx-de** [comment]
As a short proof-of-concept I've run a shell-console on my nginx-proxy-manager docker instance trying this:
```
[root@docker-nginx-ssl-proxy:/usr/local/bin]# wget https://github.com/FiloSottile/mkcert/releases/download/v1.4.1/mkcert-v1.4.1-linux-amd64

root@docker-nginx-ssl-proxy:/usr/local/bin]# mkcert *.typoworx.lan
Using the local CA at "/root/.local/share/mkcert" ✨

Created a new certificate valid for the following names 📜
 - "*.test.lan"

Reminder: X.509 wildcards only go one level deep, so this won't match a.b.typoworx.lan ℹ️

The certificate is at "./_wildcard.typoworx.lan.pem" and the key at "./_wildcard.typoworx.lan-key.pem" ✅
```

**Comment by bitsvital** [comment]
For now I use minica. It's super easy. Just spin up an ubuntu:20.04 docker. I have all the instructions written out. If you want the instructions just message me and I'll send them over to you.
https://github.com/jsha/minica

**Comment by WillJBrown** [comment]
I'd be interested in those instructions if it automates the process a little more than what typoworx-de described

**Comment by jc21** [comment]
The `mkcert` binary is shipped with the docker image, but it's not used by the software yet. I was planning to add it as an option on the SSL dropdown, but other things have taken more priority. PR's are welcome :)

**Comment by bitsvital** [comment]
> I'd be interested in those instructions if it automates the process a little more than what typoworx-de described


Hi @WillJBrown ,
I actually created a docker image that does it for you. I have all the instructions typed out in the repository. You can use the docker image or just spin up a Ubuntu image yourself and the instructions are about the same. If you run into any problems, questions, or need any help just let me know. I’ll be more than happy to help you.
Here is the the docker image.
[https://hub.docker.com/r/bitsvital/minica-bv](https://hub.docker.com/r/bitsvital/minica-bv)

**Comment by WillJBrown** [comment]
Thanks for that @bitsvital. I got it working today thanks to your page. you might like to clarify that the cert you have to share to clients is the root minica one whereas the one npm needs is the domain specific one. Also thanks to you @jc21 for npm - It's made all the local proxies I set up today so much easier. I don't know any web dev otherwise I would definitely work on a pull request to get this implemented. My knowledge is more in c#, fortran, python, etc. apologies.

**Comment by bitsvital** [comment]
@WillJBrown no problem. Thanks for the FYI. I will update that this evening. Feel free to contact me anytime if you need further assistance.

**Comment by Albonycal** [comment]
any updates on this?

**Comment by bonelifer** [comment]
Having mkcert would be awesome as some of my homelab is local only. I really don't want people being able to change the music playing in my mpd/mympd installation.

**Comment by guitaristtom** [comment]
Is there any update for this?

**Comment by onlineapps-cloud** [comment]
Any updates?

**Comment by tapionx** [comment]
It would be nice to have this!

**Comment by flavienbwk** [comment]
Indeed it might be nice. It looks like [it was possible before](https://github.com/NginxProxyManager/nginx-proxy-manager/issues/576#issuecomment-679076377). Why did you remove it ?

**Comment by flavienbwk** [comment]
Community wants to keep it open

**Comment by onlineapps-cloud** [comment]
agree with you.

**Comment by robnewport** [comment]
Please keep this open and active.

**Comment by W1BTR** [comment]
This is still a much wanted feature! Would love to see this added.

**Comment by an0o0nym** [comment]
I would also love to to see it working with NPM!

**Comment by flavienbwk** [comment]
👍

**Comment by tapionx** [comment]
yes

**Comment by huco95** [comment]
👍

**Comment by noto10** [comment]
👍

**Comment by r-nd-m** [comment]
👍

**Comment by alexunderboots** [comment]
+

**Comment by itzTheMeow** [comment]
@jc21 is this something I am able to work on? I saw elsewhere you were redoing the frontend in react, would this interfere?

**Comment by jc21** [comment]
Yeah it might interfere with the react rewrite. Given the explanation on #4525 wouldn't using a separate self-hosted CA be better? the only limitation is that this project doesn't properly support using a specific CA url in the certbot command on a per-certificate basis.

For example, in the [test stack](https://github.com/NginxProxyManager/nginx-proxy-manager/blob/develop/docker/docker-compose.ci.yml), I spin up StepCA container and PowerDNS when I request a cert for `website.example.com` it's able to issue a certificate locally. PowerDNS is only required for DNS01 certs. In order for this to work, these env vars are present:
```
LE_SERVER: 'https://ca.internal/acme/acme/directory'
REQUESTS_CA_BUNDLE: '/etc/ssl/certs/NginxProxyManager.crt'
```
If they could be defined on a per-certificate level in NPM, then you'd be set to provision your own stuff.

In my v3 prototype I was working on the principle of having [CertificateAuthorities](https://github.com/NginxProxyManager/nginx-proxy-manager/blob/v3/backend/embed/migrations/postgres/20201013035318_initial_schema.sql#L69) rows that would be [selectable when requesting certs](https://github.com/NginxProxyManager/nginx-proxy-manager/blob/v3/backend/embed/migrations/postgres/20201013035839_initial_data.sql#L65).

_A little bit off topic but I know this q is going to come up: Instead of pumping a lot of effort finishing v3 I'm trying to backport concepts back into v2 slowly until it is eventually v3. The first thing is React._

**Comment by itzTheMeow** [comment]
The certificate authority table approach seems best. Currently the way I am generating certificates is using openssl, rather than a URL or ACME server.
```bash
openssl x509 -req -in example.internal.csr -CA exampleCA.pem -CAkey exampleCA.key \
  -CAcreateserial -out example.internal.crt -days 825 -sha256 -extfile example.internal.ext
```

**Comment by Elektrik1** [comment]
2026, Still waiting for this feature!

**Comment by sduensin** [comment]
Just installed again to see if self signed certs were supported.   I need it!   :-)

**Comment by Elektrik1** [comment]
2026-06 - still waiting.

**Comment by Am0rphous** [comment]
Hi, any update on this?
